README.md
<div align="center">

# Port Scout

Port Scout is an easy way to get available ports with different needs.

</div>

<div align="center">

[![NPM version](https://img.shields.io/npm/v/port-scout.svg?style=flat)](https://www.npmjs.com/package/port-scout)
[![Known Vulnerabilities](https://snyk.io/test/github/robertcorponoi/port-scout/badge.svg)](https://snyk.io/test/github/robertcorponoi/port-scout)
[![NPM downloads](https://img.shields.io/npm/dm/port-scout.svg?style=flat)](https://www.npmjs.com/package/port-scout)
<a href="https://badge.fury.io/js/port-scout"><img src="https://img.shields.io/github/issues/robertcorponoi/port-scout.svg" alt="issues" height="18"></a>
<a href="https://badge.fury.io/js/port-scout"><img src="https://img.shields.io/github/license/robertcorponoi/port-scout.svg" alt="license" height="18"></a>
[![Gitter](https://badges.gitter.im/gitterHQ/gitter.svg)](https://gitter.im/robertcorponoi)

</div>

## **Table of contents**

- [Install](#install)
- [Usage](#usage)
- [API](#api)
- [CLI Usage](#cli-usage)

## **Install**

To install port-scout to use in your project locally, you can use:

```bash
$ npm install port-scout
```

If you plan on using port-scout as a CLI utility, you can install it globally to have access to the `port-scout` command anywhere:

```bash
$ npm install -g port-scout
```

## **Usage**

Port Scout is exported as a series of functions so you can just require the `port-scout` module and use the method you desire to retrieve the port you want.

**Note** All port scouting methods are async and so will have to be wrapped in an async function call.

```js
const scouter = require('port-scout');

async function main() {
  const webPort = await scouter.web();
}

main();
```

## **API**

### **web**

Checks common web ports to see what port is available and returns it.

**Example:**

```js
async function main() {
  const webPort = await scouter.web();
}

main();
```

### **random**

Tries random ports and returns the first available one. Ports 1-1024 are automatically avoided unless specified.

| param           	| type    	| description                                               	| default 	|
|-----------------	|---------	|-----------------------------------------------------------	|---------	|
| allowRestricted 	| boolean 	| If set to true, this will allow ports 1-1024 to be tried. 	| false   	|

**Example:**

```js
async function main() {
  const randomPort = await scouter.random();
}

main();
```

### **range**

Checks a provided range of ports for the first available port and returns it.

| param 	| type   	| description                              	| default 	|
|-------	|--------	|------------------------------------------	|---------	|
| start 	| number 	| The start of the range of ports to check 	|         	|
| end   	| number 	| The end of the range of ports to check   	| 65535   	|

**Example:**

```js
async function main() {
  const rangePort = await scouter.range(3050, 3075);
}

main();
```

## **CLI Usage**

Port can be used as a command line utility to retrieve an available port using one of its various methods and print it to the console like so:

```bash
$ port=scout -w
```

The customize the way ports are retrieved you can use a combination of the flags below:

```
-w, --web                : Returns a common web server port that can be used, these are normally in the low 3000s or 8080s.
-r, --random             : Returns a random port that can be used at the time of calling this.
-a, --range <start,end>  : Returns the first available port between a start and end value.
-f, --force              : This can be used in conjunction with the random or range flag to override the default rule of not allowing a port between 1 - 1024 to be tested.
```

**Example:**

For example, if you would like to get the first available port between 3050 and 3075 you can use:

```bash
$ port-scout -r 3050,3075
```

CHANGELOG.md
## 1.1.2 / 2020-04-16
- [MISC] Updated out-of-date dependencies to their latest versions which also fixed all fixed security vulnerabilities.
- [MISC] Changed CHANGELOG format.

## 1.1.1 / 2019-02-01
- [MISC] Updated dependencies to their latest versions.
- [MISC] Updated license year to reflect new year.

## 1.1.0 / 2019-12-15
- [FEATURE] Added 'num-step' package and used it in the 'web' method to alternate between ports 3000, 8000, and 8080 when searching for an available web port.
- [MISC] Added repository field to package.json and updated existing information.

## 0.1.1 / 2019-11-05
- [MISC] Updated dependencies to their latest versions.
- [MISC] Updated README.

## 0.1.0 / 2019-09-17
- Initial release


test/port-scout.test.js
'use strict'

const net = require('net');
const chai = require('chai');
const scanner = require('../dist/index');

const servers = [];

afterEach(async () => await clearPorts());

describe('Finding available web ports', () => {

  it('should find port 3000 as the first available web port', async () => {

    const port = await scanner.web();

    chai.expect(port).to.equal(3000);

  });

  it('should find port 8080 as the first available web port as 3000 and 8000 are taken by a test process', async () => {

    occupyPort(3000);
    occupyPort(8000);

    const port = await scanner.web();

    clearPorts();

    chai.expect(port).to.equal(8080);

  });

  it('should find port 3001 as the first available web port as 3000, 8000, and 8080 are taken by a test process', async () => {

    occupyPort(3000);
    occupyPort(8000);
    occupyPort(8080);

    const port = await scanner.web();

    clearPorts();

    chai.expect(port).to.equal(3001);

  });

});

describe('Finding available random ports', () => {

  it('should find an available random port', async () => {

    const port = await scanner.random();

    chai.expect(port).to.be.greaterThan(1024) && chai.expect(port).to.be.lessThan(65536);

  });

  it('should find an available random port without restrictions', async () => {

    const port = await scanner.random();

    chai.expect(port).to.be.greaterThan(0) && chai.expect(port).to.be.lessThan(65536);

  });

});

describe('Finding available ports in a range', () => {

  it('should find an available port between 4000 and 4010', async () => {

    const port = await scanner.range(4000, 4010);

    chai.expect(port).to.be.greaterThan(3999) && chai.expect(port).to.be.lessThan(4011);

  });

  it('should find an available port between 4000 and the default end of 65535', async () => {

    const port = await scanner.range(4000);

    chai.expect(port).to.be.greaterThan(3999) && chai.expect(port).to.be.lessThan(65536);

  });

});


/**
 * Makes a port unavailable for testing purposes.
 * 
 * @param {number} port The port that should become unavailable.
 */
async function occupyPort(port) {

  const server = net.createServer();

  servers.push(server);

  await server.listen(port);

}

/**
 * Clears all testing ports and makes them available again.
 */
async function clearPorts() {

  servers.map(async server => {

    await server.close();

  });

}

src/index.ts
'use strict'

import * as net from 'net';
// @ts-ignore
import Stepper from 'num-step';
import { randomInteger } from './utils';

/**
 * Finds the first available port that should be used based on conditions specified by the user.
 */
module.exports = {

  /**
   * Checks common web ports to see what is available.
   * 
   * @async
   * 
   * @returns {number} Returns the first available web port.
   */
  async web(): Promise<number> {

    const stepper = new Stepper([3000, 8000, 8080]);

    let port: number = 3000;

    let amountToStep: number = 1;

    let success: boolean = false;

    while (!success) {

      try {

        success = await checkPort(port);

        return port;

      } catch (err) {

        amountToStep++;

        port = stepper.step(amountToStep);

      }

    }

  },

  /**
   * Tries random ports and returns the first available one.
   * 
   * @async
   * 
   * @param {boolean} [allowRestricted=false] If set to true, this will allow ports under 1024 to be tried.
   * 
   * @returns {number} Returns the first available random port.
   */
  async random(allowRestricted: boolean = false): Promise<number> {

    const min = allowRestricted ? 1 : 1025;
    const max = 65535;

    let port = randomInteger(min, max);

    let success: boolean = false;

    while (!success) {

      try {

        success = await checkPort(port);

        return port;

      } catch (err) {

        port = randomInteger(min, max);


      }

    }

    return port;

  },

  /**
   * Checks a provided range of ports for an available port.
   * 
   * @async
   * 
   * @param {number} start The start of the range of ports to check.
   * @param {number} [end=65535] The end of the range of ports to check.
   * 
   * @returns {number} Returns the first available port in the provided range.
   */
  async range(start: number, end: number = 65535): Promise<number> {

    let port: number = start;

    let success: boolean = false;

    while (!success) {

      try {

        success = await checkPort(port);

        if (port > end) throw new Error('Unable to find an available port in the specified range');

        return port;

      } catch (err) {

        port++;


      }

    }

    return port;

  },

}

/**
 * Checks to see what the first available port is according to the specified options.
 * 
 * @async
 * @private
 * 
 * @param {number} port The port number to check if available or not.
 * 
 * @returns {boolean}
 */
function checkPort(port: number): Promise < boolean > {

  return new Promise((resolve, reject) => {

    const server = net.createServer();

    server.once('error', (err: any) => {

      if (err.code === 'EADDRINUSE') reject(err);

    });

    server.once('listening', () => {

      server.close();

      resolve(true);

    });

    server.listen(port);

  });

}